### PR TITLE
Fix(#341): Remove deadlock when rebalancing by multiple consumer instances

### PR DESCRIPTION
**Pull request by merrygoround-of-life**
this is to fix the issue: #341
when rebalancing starts by multiple instances in a consumer group, a pseudo deadlock condition occurs:

thread a. (Consumer Expiration Thread)
 1) acquired the KafkaConsumerManager instance (lock A) within KafkaConsumerManager$ExpirationThread::Run
 2) waiting to acquire the KafkaConsumerState instance (lock B) within KafkaConsumerState::expired
thread-a-monitor

thread b.
 1) acquired the KafkaConsumerState instance (lock B) within KafkaConsumerState::hasNext in KafkaConsumerReadTask::addRecords
 2) waiting to acquire the ?? (lock C) within KafkaConsumer::poll(0) by rebalancing
thread-b-running

thread c.
 1) acquired the ?? (lock C) within jettyHandler
 2) waiting to acquire the KafkaConsumerManager instance (lock A) within KafkaConsumerManager::getConsumerInstance in readRecords
thread-c-monitor

a cyclic lock order is generated by 3 locks and 3 threads, resulting in deadlock.
to resolve this, split the KafkaConsumerState instance lock (lock B) into two locks - a KafkaConsumerState instance lock(existing) and KafkaConsumerState.expiration member lock(new).
as a result, lock B of thread a and lock B of thread b become different locks, breaking the cycle.

since thread b is actually in the running state, the lock C is not exactly the the Java synchronization object.
perhaps it needs permission(lock) to use jetty's handler when certain events are triggered by rebalancing inside KafkaConsumer::poll, which I've referred to as the lock C.

### CODE REVIEW

File: kafka-rest/src/main/java/io/confluent/kafkarest/v2/KafkaConsumerState.java
Comment (plantrue) [inline kafka-rest/src/main/java/io/confluent/kafkarest/v2/KafkaConsumerState.java:69]: Added a separate final object for synchronizing the expiration field.

File: kafka-rest/src/main/java/io/confluent/kafkarest/v2/KafkaConsumerState.java
Comment (plantrue) [inline kafka-rest/src/main/java/io/confluent/kafkarest/v2/KafkaConsumerState.java:347]: Resolve a deadlock by making the scope of synchronized small.
(replacing the KafkaConsumerState instance lock with a new KafkaConsumerState::expirationLock instance member lock)

**Comment by klboke** [comment]
@merrygoround-of-life  Hi, what a coincidence, I also found this problem yesterday, my jstack analysis results are somewhat similar to your exposition, but not exactly the same. Finally, did you solve your problem?

**Comment by merrygoround-of-life** [comment]
@klboke the stack information may be slightly different depending on the kafka-rest version you use. but i think it would be the same problem i found if you could see a thread waiting indefinitely for the KafkaConsumer::poll function to return.
Finally, it was very difficult to reproduce that issue with this modification.

**Comment by klboke** [comment]
@merrygoround-of-life  It should be the same problem, this is the stack information on my side, eventually all waiting for the poll operation, while the poll operation is waiting for rebalancing
image

**Comment by merrygoround-of-life** [comment]
according to my analysis, a deadlock was caused by 3 or more locks and threads, but it was not clear who the C lock was.
- thread a: lock A -> lock B
- thread b: lock B -> lock C
- thread c: lock C -> lock A
In your diagram, it appears that lock A (c22fa4d0) corresponds to a KafkaConsumerManager singleton instance and lock B (c3bdf508) corresponds to a KafkaConsumerState instance. (A KafkaConsumerState instance is created for each Consumer instance) however it's still not clear from your diagram who the third lock that completes the circular waiting is.

Presumably some of the threads called getConsumerInstance function are expected to be in the context of request handler of the jetty web server and it is presumed that the thread waiting for the return of KafkaConsumer::poll(0) in addRecords function need to acquire some lock associated with the jetty handler for rebalancing which I've referred to as the lock C. but no matter who lock C is, my solution is to split the lock B to break the circular waiting.

what version of kafka-rest have you tested?
I'll check to see if this solution is equally valid in the kafka-rest of your version,

**Comment by merrygoround-of-life** [comment]
it seems to be related to issues #789 and #587 as well.

**Comment by klboke** [comment]
I'm using the https://github.com/confluentinc/kafka-rest/tree/6.2.1-post version. Your analysis is correct, but it seems that there is no lock on C. All thread operations end up being blocked by the addRecords lock, which makes it impossible for the thread to clean up the expired consumer. The following diagram.
image
addRecords holds a hasNext lock, blocking getConsumerInstance, blocking ExpirationThread.run

**Comment by merrygoround-of-life** [comment]
@klboke I confirmed that in my use case the deadlock was equally reproduced in the 6.2.1-post version and it was not reproduced when this modification was applied.
Can you check if it is solved in your use case by applying the attached patch file?
- [patch.txt](https://github.com/confluentinc/kafka-rest/files/7821000/patch.txt)
> patch -p1 < patch.txt

**Comment by merrygoround-of-life** [comment]
@klboke I confirmed that in my use case the deadlock was equally reproduced in the 6.2.1-post version and it was not reproduced when this modification was applied.
Can you check if it is solved in your use case by applying the attached patch file?
- [patch.txt](https://github.com/confluentinc/kafka-rest/files/7821103/patch.txt)
> patch -p1 < patch.txt

**Comment by klboke** [comment]
It should be solved, as I posted above, according to your pr so modified, ExpirationThread.run will be able to run properly, I think the key point is here

**Comment by merrygoround-of-life** [comment]
@klboke
the hypothesis of my analysis is that 'KafkaConsumer::poll function is not returning because a deadlock has occurred'.
it is known that a deadlock occurs when multiple threads need the same locks but obtain them in different order,
so if there are only two locks that caused the deadlock, you should be able to see the A -> B lock acquisition order and vice versa.
however in our analysis, we found an A -> B lock acquisition order but not a reverse acquisition order (B -> A).

if there is another undiscovered lock (C) associated with the deadlock, it can be assumed that the lock acquisition order of B -> (C) -> A exists (though not explicitly seen) and thus the deadlock is established.

148019839-49e986ee-eb28-46f3-b5ea-62990ffbb948

A scenario is possible that the KafkaConsumer::poll function is not returning because the KafkaConsumer::poll function is trying to acquire the lock C internally, but has already been acquired by another thread (finally calling the KafkaConsumer::poll function).
i'm guessing this has something to do with the issues found when searching for 'jetty deadlock'.

**Comment by klboke** [comment]
@merrygoround-of-life 
To be sure, there is no deadlock found in my program stack. If there was a deadlock, it would print [Found 1 deadlock] in the jstack result, however, there is not, and this is also true from the analysis of the program stack. All the blocking ends up waiting for KafkaConsumer::poll, ExpirationThread is also blocked, and updating the expiration time is also blocked, here is the key point

**Comment by merrygoround-of-life** [comment]
@klboke you are right. this issue is not a ‘real’ deadlock consisting of only java sync objects. that's why i used the term 'pseudo deadlock' (or logical deadlock).
i don't think it can explain why the KafkaConsumer::poll function doesn't block when i split the lock B, unless it’s a deadlock (circular wait).

File: kafka-rest/src/main/java/io/confluent/kafkarest/v2/KafkaConsumerState.java
Comment (rigelbm) [reply kafka-rest/src/main/java/io/confluent/kafkarest/v2/KafkaConsumerState.java:69]: Can you just do `Object expirationLock = new Object()` instead? We don't actually need a `Long` for anything.

Comment (merrygoround-of-life) [reply kafka-rest/src/main/java/io/confluent/kafkarest/v2/KafkaConsumerState.java:69]: @rigelbm No problem. I'll fix it soon based on your comment.

**Review by rigelbm — APPROVED** [review]
LGTM

**Comment by merrygoround-of-life** [comment]
@vharitonsky 'Only those with write access to this repository can merge pull requests.'
@rigelbm Could you by any chance merge this? or is there a guide to do this own my own?

**Comment by vharitonsky** [comment]
Guys, this merge is very important for our infrastructure, if you could maybe look into it?

**Comment by ehumber** [comment]
@vharitonsky  I'm looking at this at the moment - aiming to merge it very soon.

**Review by ehumber — APPROVED** [review]
Have tested locally and reviewed the changes.  All looks good to me.

**Comment by ehumber** [comment]
@merrygoround-of-life Thanks for solving this, and delivering the code fix, really appreciate you to doing it.

I've merged your change into 6.0.x, and also merged it into all later release branches.

**Comment by techexplorer0310** [comment]
Hi, 

In which version should we get the fix?

Thanks,

**Pull request merged**
